File: fribidi-vs-unicode/algo_n0.py
#!/usr/bin/python

import bd16
import pandas as pd

# Identify the bracket pairs in the current isolating run seq
#pairings = bd16.find_pairings(text)

def is_strong(bc):
  return bc in ['R','L']

def get_bidi_class(ch):
  if ch in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ':
    return 'R'
  if ch in 'abcdefghijklmnopqrstuvwxyz':
    return 'L'
  if ch in '({[]})&!':
    return 'ON'
  if ch in ' ':
    return 'WS'

def find_strong(bidi_classes, start, end):
  '''find a strong character between start and end non-inclusive'''
  rdir = 1
  if end < start:
    if end == -1:
      end = None   # Req of the string range operator
    rdir = -1

  for bc in bidi_classes[start:end:rdir]:
    if is_strong(bc):
      return bc

  return None

def find_matching_strong(bidi_classes, start, end, match):
  '''Used to find a character between start and end with the
  bidiclass match'''
  for i in range(start,end):
    if bidi_classes[i] == match:
      return True

  return False

def test_n0(text,embedding_direction, print_pairings=False):
  Table = [list(text) + ['']]
  bc = [get_bidi_class(ch) for ch in text]
  Table += [bc[:] + ['']]
  
  pairings = bd16.find_pairings(text)
  if print_pairings:
    print pairings
  for pr in pairings:
    rule = ''
    other_direction = {'L':'R','R':'L'}[embedding_direction]
    has_matching_strong = find_matching_strong(bc, pr[0], pr[1], embedding_direction)
    has_other_strong = find_matching_strong(bc, pr[0], pr[1], 
                                            other_direction)
    preceding_strong = find_strong(bc, pr[0], -1)
    print '(%d,%d) has_matching_strong has_other_strong preceding_strong='%(pr[0],pr[1]),has_matching_strong,has_other_strong, preceding_strong
  
    # Rule N0b
    if has_matching_strong:
      rule= 'N0b: %s->%s'%(bc[pr[0]],embedding_direction)
      bc[pr[0]]=bc[pr[1]]=embedding_direction
  
    elif has_other_strong:
      # Rule N0c1
      if preceding_strong == other_direction:
        rule= 'N0c1: %s->%s'%(bc[pr[0]],other_direction)
        bc[pr[0]]=bc[pr[1]]=other_direction
      else:
        rule= 'N0c2: %s->%s'%(bc[pr[0]],embedding_direction)
        bc[pr[0]]=bc[pr[1]]=embedding_direction
    else:
      rule= 'N0d'
      # N0d - Do nothing
  
    Table += [bc[:]+[rule,pr]]
  df =  pd.DataFrame(Table,
                     columns=range(len(Table[0])-1)+['Rule','Pair'])
  df.Pair.fillna(value='',inplace=True)
  print df
  print ''
  

#test_n0(text='AB(CD[&ef]!)gh',embedding_direction='R', print_pairings=True)
#test_n0(text='sm (fa AR) HE',embedding_direction='R')
#test_n0(text='AR bo(s)',embedding_direction='R')
#test_n0(text='(a)(b)c',embedding_direction='R', print_pairings=True)
test_n0(text='a(b)).A',embedding_direction='R', print_pairings=True)

File: fribidi-vs-unicode/algo-n0.c
// A straight translation of algo_n0.py to C.

#include <glib.h>
#include <string.h>
#include <stdio.h>
#include "algo-n0.h"
#include "bd16.h"

gboolean charinstring(gchar ch, const char *string)
{
  const char *p=string;
  while(*p)
      if (*p++ == ch)
        return TRUE;
  return FALSE;
}

gboolean is_strong(bidi_class_t bc)
{
  return bc=='R' || bc=='L';
}

bidi_class_t get_bidi_class(gchar ch)
{
  if (charinstring(ch, "ABCDEFGHIJKLMNOPQRSTUVWXYZ"))
    return (bidi_class_t)'R';
  if (charinstring(ch, "abcdefghijklmnopqrstuvwxyz"))
    return (bidi_class_t)'L';
  if (charinstring(ch, "({[]})&!"))
    return (bidi_class_t)'N';
  if (charinstring(ch, " "))
    return (bidi_class_t)'N';
  return (bidi_class_t)'?';
}

// Find a strong character between start and end (non-inclusive)
bidi_class_t find_strong(bidi_class_t *bidi_classes, int start, int end)
{
  int dir= end < start ? -1 : 1;
  int i;
  
  for (i=start; i!=end; i+= dir)
    {
      if (is_strong(bidi_classes[i]))
        return bidi_classes[i];
    }
  return (bidi_class_t)'?';
}

// Used to find a character beetween start and end with the matching bidi class
gboolean find_matching_strong(bidi_class_t *bidi_classes, int start, int end, bidi_class_t match)
{
  int i;
  for (i=start; i<end; i++)
    if (bidi_classes[i] == match)
      return TRUE;
  return FALSE;
}

void test_n0(gchar *text, bidi_class_t embedding_direction)
{
  int n = strlen(text);
  int i;
  for (i=0; i<n; i++)
      printf("%2d ",i);
  printf("\n");
  for (i=0; i<n; i++)
      printf("%2c ",text[i]);
  printf("\n");
    

  pairing_node_t *pairings = find_pairings(text);

  bidi_class_t *bidi_classes = g_new0(bidi_class_t, n);
  for (i=0; i<n; i++)
    bidi_classes[i] = get_bidi_class(text[i]);
  for (i=0; i<n; i++)
      printf("%2c ",bidi_classes[i]);
  printf("\n");

  pairing_node_t *pr = pairings;
  while(pr)
    {
      char *rule = NULL;
      gboolean has_strong = find_matching_strong(bidi_classes,
                                                 pr->start, pr->end,
                                                 embedding_direction);
      bidi_class_t preceding_strong = find_strong(bidi_classes,
                                                  pr->start, -1);

      // Rule N0b
      if (has_strong)
        {
          rule= g_strdup_printf("N0b: %c->%c",bidi_classes[pr->start],embedding_direction);
          bidi_classes[pr->start]=bidi_classes[pr->end]=embedding_direction;
        }

      else if (charinstring(preceding_strong,"RL"))
        {
          // Rule N0c1
          // printf("preceding_strong embedding_direction=%c %c",
          //        preceding_strong,embedding_direction);
          if (preceding_strong != embedding_direction)
            {
              rule= g_strdup_printf("N0c1: %c->%c",
                                    bidi_classes[pr->start],
                                    preceding_strong);
              bidi_classes[pr->start]=bidi_classes[pr->end]=preceding_strong;
            }
          // Rule N0c2
          else
            {
              rule= g_strdup_printf("N0c2: %c->%c",
                                    bidi_classes[pr->start],embedding_direction);
              bidi_classes[pr->start]=bidi_classes[pr->end]=embedding_direction;
            }
        }
      else
        {
          // N0d - Do nothing;
          rule = g_strdup("N0d");
        }
      for (i=0; i<n; i++)
        printf("%2c ",bidi_classes[i]);
      printf("%s\n", rule);
      g_free(rule);
      pr = pr->next;
    }

  g_free(bidi_classes);
  free_parings(pairings);
  printf("\n");
}

